jupyterhub/templates/proxy/autohttps/configmap.yaml

```diff
@@ -1,134 +1,28 @@
 {{- $HTTPS := (and .Values.proxy.https.hosts .Values.proxy.https.enabled) }}
 {{- $autoHTTPS := (and $HTTPS (eq .Values.proxy.https.type "letsencrypt")) }}
 {{- if $autoHTTPS -}}
+{{- $_ := required .Values.proxy.https.letsencrypt.contactEmail "proxy.https.letsencrypt.contactEmail is a required field" -}}
+
+# This configmap contains Traefik configuration files to be mounted.
+# - traefik.yaml will only be read during startup (static configuration)
+# - dynamic.yaml will be read on change (dynamic configuration)
+#
+# ref: https://docs.traefik.io/getting-started/configuration-overview/
+#
+# The configuration files are first rendered with Helm templating to large YAML
+# strings. Then we use the fromYAML function on these strings to get an object,
+# that we in turn merge with user provided extra configuration.
+#
 kind: ConfigMap
 apiVersion: v1
 metadata:
   name: traefik-proxy-config
   labels:
     {{- include "jupyterhub.labels" . | nindent 4 }}
 data:
-  # This configmap provides the complete configuration for our traefik proxy.
-  # traefik has 'static' config and 'dynamic' config (https://docs.traefik.io/getting-started/configuration-overview/)
-  # traefik.toml contains the static config, while dynamic.toml has the dynamic config.
-  traefik.toml: |
+  traefik.yaml: |
+    {{- include "jupyterhub.traefik.yaml" . | fromYaml | merge .Values.proxy.traefik.extraStaticConfig | toYaml | nindent 4 }}
```

Comment:
What happens if the extraStaticConfig redefines some of the config already set in `jupyterhub.traefik.yaml`? Will those be replaced?

Reply:
Yepp! extraStaticConfig will have the final say through a merge. To delete something one can specify something to be null rather than leaving it out.

Reply:
To merge like this is often problematic when working with lists, because how to merge for example [1,2,3] with [4], replace or extend? Both could make sense, but in this case, it will be a replace operation, and if we wanted to extend, it wouldn't be possible.

Thankfully, lots of the configuration is not using elements in lists, but instead keys with arbitrary names in dictionaries such as the middleware and entrypoint names.